common/types.go

```diff
@@ -178,25 +178,57 @@ type RegistryCredential struct {
 	BasicAuth BasicAuthCredential `json:"basicauth,omitempty"`
 }
 
+// Registry describes a template registry
+// TODO(jackr): Fix ambiguity re: whether or not URL has a scheme.
 type Registry struct {
-	Name       string             `json:"name,omitempty"` // Friendly name for the repo
-	Type       RegistryType       `json:"type,omitempty"` // Technology implementing the registry
-	URL        string             `json:"name,omitempty"` // URL to the root of the repo, for example: github.com/helm/charts
+	Name   string         `json:"name,omitempty"`   // Friendly name for the registry
+	Type   RegistryType   `json:"type,omitempty"`   // Technology implementing the registry
+	URL    string         `json:"name,omitempty"`   // URL to the root of the registry
+	Format RegistryFormat `json:"format,omitempty"` // Format of the registry
+}
+
+// AuthenticatedRegistry describes a type registry with credentials.
+type AuthenticatedRegistry struct {
+	Registry
 	Credential RegistryCredential `json:"credential,omitempty"`
-	Format     RegistryFormat     `json:"format,omitempty"`
 }
 
 // RegistryType defines the technology that implements the registry
 type RegistryType string
 
 const (
-	Github RegistryType = "github"
+	GithubRegistryType RegistryType = "github"
 )
 
-// RegistryFormat defines the format of the registry
+// RegistryFormat is a semi-colon delimited string that describes the format
+// of a registry.
 type RegistryFormat string
 
 const (
-	VersionedRegistry   RegistryFormat = "versioned"
+	// Versioning.
+	// If a registry if versioned, then types appear under versions.
+	VersionedRegistry RegistryFormat = "versioned"
+	// If a registry is unversioned, then types appear under their names.
 	UnversionedRegistry RegistryFormat = "unversioned"
+
+	// Organization.
+	// In a collection registry, types are grouped into collections.
+	CollectionRegistry RegistryFormat = "collection"
```

Comment:
Thanks for catching that.

Re: expected format, things have been pretty ad hoc, up to this point. For example, we hard wire the names of kubernetes/application-dm-templates and helm/charts. Doesn't work well with the notion of letting the user manage a list of registries.

Likewise, we assume that all versioned registries have collections, that all unversioned registries don't, and that we won't see any other registry layouts. As this functionality evolves, we will almost certainly end up making changes to the registry layout.

Likewise, we assume in some places that the registry URL doesn't include a scheme (i.e. https://), but in other places, we assume it does. We haven't been bitten by this one yet, but the struct and interface definitions we currently have don't differentiate clearly, so we will be bitten before long.

The goal of the format string change, as well as most of the others in this PR, is to unwind those assumptions and others like them, and to expose a disciplined framework for working with registries that makes assumptions explicit and that's easy to maintain.

I toyed with the idea of replacing the format string with a general purpose map containing registry metadata (i.e., key/value pairs), which we could load from a metadata file at the root of the registry. I didn't have the appetite to go that far in this pass, but we're going to need it, if this effort is successful.

Note that there's a difference between providing a way to define orthogonal registry properties and supporting all possible combinations. We can always return an error if we encounter an unsupported combination, or put an issue in the backlog go build support for it, if that's what our customers need.

The key is being able to express capabilities clearly, so that we can determine unambiguously what is and isn't supported in any given release.